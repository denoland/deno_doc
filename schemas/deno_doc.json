{
  "$id": "https://deno.land/schemas/deno_doc.json",
  "$schema": "http://json-schema.org/draft-07/schema",
  "description": "A JSON representation of a deno_doc documentation structure",
  "type": "array",
  "items": {
    "$ref": "#/definitions/DocNode"
  },
  "definitions": {
    "DocNode": {
      "type": "object",
      "required": [
        "kind",
        "name",
        "location"
      ],
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "function",
            "variable",
            "enum",
            "class",
            "typeAlias",
            "namespace",
            "interface",
            "import"
          ]
        },
        "name": {
          "type": "string"
        },
        "location": {
          "$ref": "#/definitions/Location"
        },
        "jsDoc": {
          "type": "string"
        }
      },
      "anyOf": [
        {
          "required": [
            "functionDef"
          ],
          "properties": {
            "kind": {
              "const": "function"
            },
            "functionDef": {
              "$ref": "#/definitions/FunctionDef"
            }
          }
        },
        {
          "required": [
            "variableDef"
          ],
          "properties": {
            "kind": {
              "const": "variable"
            },
            "variableDef": {
              "$ref": "#/definitions/VariableDef"
            }
          }
        },
        {
          "required": [
            "enumDef"
          ],
          "properties": {
            "kind": {
              "const": "enum"
            },
            "enumDef": {
              "$ref": "#/definitions/EnumDef"
            }
          }
        },
        {
          "required": [
            "classDef"
          ],
          "properties": {
            "kind": {
              "const": "class"
            }
          },
          "classDef": {
            "$ref": "#/definitions/ClassDef"
          }
        },
        {
          "required": [
            "typeAliasDef"
          ],
          "properties": {
            "kind": {
              "const": "typeAlias"
            }
          },
          "typeAliasDef": {
            "$ref": "#/definitions/TypeAliasDef"
          }
        },
        {
          "required": [
            "namespaceDef"
          ],
          "properties": {
            "kind": {
              "const": "namespace"
            }
          },
          "namespaceDef": {
            "$ref": "#/definitions/NamespaceDef"
          }
        },
        {
          "required": [
            "interfaceDef"
          ],
          "properties": {
            "kind": {
              "const": "interface"
            }
          },
          "interfaceDef": {
            "$ref": "#/definitions/interfaceDef"
          }
        },
        {
          "required": [
            "importDef"
          ],
          "properties": {
            "kind": {
              "const": "import"
            },
            "importDef": {
              "$ref": "#/definitions/ImportDef"
            }
          }
        }
      ]
    },
    "Location": {
      "type": "object",
      "required": [
        "filename",
        "line",
        "col"
      ],
      "properties": {
        "filename": {
          "type": "string"
        },
        "line": {
          "type": "number"
        },
        "col": {
          "type": "number"
        }
      }
    }
  }
}
