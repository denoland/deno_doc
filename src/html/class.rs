use super::util::*;
use crate::html::parameters::render_params;
use std::fmt::Write;

pub fn render_class(doc_node: &crate::DocNode, ctx: &RenderContext) -> String {
  let class_def = doc_node.class_def.as_ref().unwrap();

  let ctx = &RenderContext {
    current_type_params: class_def
      .type_params
      .iter()
      .map(|def| def.name.clone())
      .collect::<std::collections::HashSet<String>>(),
    ..ctx.clone()
  };

  // TODO: class items

  format!(
    r#"<div class="doc_block_items">{}{}{}{}</div>"#,
    super::jsdoc::render_docs(&doc_node.js_doc, true, false, ctx),
    render_constructors(&class_def.constructors, &doc_node.name, ctx),
    super::types::render_type_params(&class_def.type_params, ctx),
    render_index_signatures(&class_def.index_signatures, ctx),
  )
}

fn render_constructors(
  constructors: &[crate::class::ClassConstructorDef],
  name: &str,
  ctx: &RenderContext,
) -> String {
  if constructors.is_empty() {
    return String::new();
  }

  let items = constructors
    .iter()
    .enumerate()
    .map(|(i, constructor)| {
      let id = name_to_id("constructor", &i.to_string());

      // TODO: tags, render constructor params
      doc_entry(&id, name, "()", constructor.js_doc.doc.as_deref(), ctx)
    })
    .collect::<String>();

  section("Constructors", &items)
}

fn render_index_signatures(
  index_signatures: &[crate::class::ClassIndexSignatureDef],
  ctx: &RenderContext,
) -> String {
  if index_signatures.is_empty() {
    return String::new();
  }

  let items = index_signatures.iter().enumerate().fold(
    String::new(),
    |mut output, (i, index_signature)| {
      let id = name_to_id("index_signature", &i.to_string());

      let readonly = index_signature
        .readonly
        .then_some("<span>readonly </span>")
        .unwrap_or_default();

      let ts_type = index_signature
        .ts_type
        .as_ref()
        .map(|ts_type| {
          format!(": {}", super::types::render_type_def(ts_type, ctx))
        })
        .unwrap_or_default();

      write!(
        output,
        r#"<div class="doc_item" id="{id}">{}{readonly}[{}]{ts_type}</div>"#,
        anchor(&id),
        render_params(&index_signature.params, ctx),
      )
      .unwrap();
      output
    },
  );

  section("Index Signatures", &items)
}
