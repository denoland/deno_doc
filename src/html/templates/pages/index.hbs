{{~> pages/html_head html_head_ctx ~}}
{{~> category_panel categories_panel ~}}
<div>
  {{~> pages/top_nav ~}}

  <div id="content">
    <main>
      {{~#if usage~}}
        {{~> usages_large usage ~}}
      {{~/if~}}

      {{~#if module_doc~}}
        {{~> module_doc module_doc ~}}
      {{~/if~}}

      {{~#if overview~}}
        {{~> symbol_content overview ~}}
      {{~/if~}}
    </main>

    {{~> toc toc_ctx ~}}

    {{~> pages/search_results ~}}
  </div>
</div>
</div>
</body>
</html>
