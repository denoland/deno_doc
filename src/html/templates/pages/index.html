{{ call pages/html_head.html with html_head_ctx }}

{{ call index_sidepanel.html with sidepanel_ctx }}

<div id="content" class="with_sidepanel">
  {{ call breadcrumbs.html with breadcrumbs_ctx }}
  {{ call pages/search_bar.html with search_ctx }}

  <main>
    {#- TODO(bartlomieju): call into template here #}

    {{ if module_doc }}
      {{ call module_doc.html with module_doc }}
    {{ endif }}
  </main>

  {{ call pages/search_results.html with search_ctx }}
<div>
</body>
</html>
