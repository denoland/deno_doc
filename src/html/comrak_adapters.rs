// Copied and modified from https://github.com/kivikakk/comrak/blob/main/src/plugins/syntect.rs

//! Adapter for the Syntect syntax highlighter plugin.

use crate::html::ShortPath;
use comrak::adapters::HeadingAdapter;
use comrak::adapters::HeadingMeta;
use comrak::adapters::SyntaxHighlighterAdapter;
use comrak::nodes::Sourcepos;
use std::collections::HashMap;
use std::io::Write;
use std::sync::Arc;
use std::sync::Mutex;

#[derive(Debug)]
/// Syntect syntax highlighter plugin.
pub struct HighlightAdapter {
  #[cfg(feature = "syntect")]
  pub syntax_set: syntect::parsing::SyntaxSet,
  #[cfg(feature = "syntect")]
  pub theme_set: syntect::highlighting::ThemeSet,
  #[cfg(feature = "tree-sitter")]
  pub language_cb:
    fn(&str) -> Option<&'static tree_sitter_highlight::HighlightConfiguration>,
  pub show_line_numbers: bool,
}

impl HighlightAdapter {
  fn highlight_html<'a, I, H>(
    &self,
    iter: I,
    mut highlighter: H,
  ) -> Result<String, anyhow::Error>
  where
    I: Iterator<Item = &'a str>,
    H: FnMut(&mut String, &str) -> Result<(), anyhow::Error>,
  {
    let mut line_numbers = String::new();
    let mut lines = String::new();

    for (i, line) in iter.enumerate() {
      let n = i + 1;

      if self.show_line_numbers {
        line_numbers.push_str(&format!(
          r##"<a href="#L{n}" class="no_color block">{n}</a>"##,
        ));

        lines.push_str(&format!(
          r#"<span id="L{n}" class="block target:bg-yellow-200">"#
        ));
      }

      highlighter(&mut lines, line)?;

      if self.show_line_numbers {
        lines.push_str("</span>");
      }
    }

    if self.show_line_numbers {
      Ok(format!(
        r##"<div class="lineNumbers">{line_numbers}</div><div class="grow overflow-x-auto">{lines}</div>"##
      ))
    } else {
      Ok(lines)
    }
  }

  fn write_button(
    &self,
    output: &mut dyn Write,
    source: &str,
  ) -> std::io::Result<()> {
    write!(output, "</code>")?;
    write!(
      output,
      r#"<button class="context_button" data-copy="{}">{}</button>"#,
      html_escape::encode_double_quoted_attribute(source),
      include_str!("./templates/icons/copy.svg")
    )?;
    write!(output, "<code>")
  }
}

impl SyntaxHighlighterAdapter for HighlightAdapter {
  #[cfg(any(
    all(feature = "syntect", not(feature = "tree-sitter")),
    all(feature = "syntect", feature = "tree-sitter")
  ))]
  fn write_highlighted(
    &self,
    output: &mut dyn Write,
    lang: Option<&str>,
    code: &str,
  ) -> std::io::Result<()> {
    let lang = match lang {
      Some(l) if !l.is_empty() => l,
      _ => "Plain Text",
    };

    let syntax =
      self
        .syntax_set
        .find_syntax_by_token(lang)
        .unwrap_or_else(|| {
          self
            .syntax_set
            .find_syntax_by_first_line(code)
            .unwrap_or_else(|| self.syntax_set.find_syntax_plain_text())
        });

    let theme = &self.theme_set.themes["InspiredGitHub"];
    let mut highlighter = syntect::easy::HighlightLines::new(syntax, theme);

    match self.highlight_html(
      syntect::util::LinesWithEndings::from(&code),
      |lines, line| {
        let regions = highlighter.highlight_line(line, &self.syntax_set)?;
        syntect::html::append_highlighted_html_for_styled_line(
          &regions[..],
          syntect::html::IncludeBackground::No,
          lines,
        )?;

        Ok(())
      },
    ) {
      Ok(highlighted_code) => output.write_all(highlighted_code.as_bytes())?,
      Err(_) => output.write_all(code.as_bytes())?,
    }

    self.write_button(output, &code)
  }

  #[cfg(all(feature = "tree-sitter", not(feature = "syntect")))]
  fn write_highlighted(
    &self,
    output: &mut dyn Write,
    lang: Option<&str>,
    code: &str,
  ) -> std::io::Result<()> {
    let lang = lang.unwrap_or_default();
    let config = (self.language_cb)(lang);
    let source = code.as_bytes();
    if let Some(config) = config {
      let mut highlighter = tree_sitter_highlight::Highlighter::new();
      let res = highlighter.highlight(config, source, None, self.language_cb);
      match res {
        Ok(highlighter) => {
          let mut renderer = tree_sitter_highlight::HtmlRenderer::new();
          match renderer.render(highlighter, source, &|highlight| {
            crate::html::tree_sitter::classes(highlight)
          }) {
            Ok(()) => {
              let html = self
                .highlight_html(renderer.lines(), |lines, line| {
                  lines.push_str(line);
                  Ok(())
                })
                .unwrap();

              output.write_all(html.as_bytes())?;
              return self.write_button(output, &code);
            }
            Err(err) => {
              eprintln!("Error rendering code: {}", err);
            }
          };
        }
        Err(err) => {
          eprintln!("Error highlighting code: {}", err);
        }
      }
    }
    comrak::html::escape(output, source)?;
    self.write_button(output, &code)
  }

  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  fn write_highlighted(
    &self,
    output: &mut dyn Write,
    _lang: Option<&str>,
    code: &str,
  ) -> std::io::Result<()> {
    let code = html_escape::encode_text(code);
    let html = self
      .highlight_html(code.lines(), |lines, line| {
        lines.push_str(&format!("{line}\n"));

        Ok(())
      })
      .unwrap();

    output.write_all(html.as_bytes())?;
    self.write_button(output, &code)
  }

  fn write_pre_tag(
    &self,
    output: &mut dyn Write,
    mut attributes: HashMap<String, String>,
  ) -> std::io::Result<()> {
    attributes
      .entry("class".into())
      .or_default()
      .push_str(" highlight");
    comrak::html::write_opening_tag(output, "pre", attributes)
  }

  fn write_code_tag(
    &self,
    output: &mut dyn Write,
    mut attributes: HashMap<String, String>,
  ) -> std::io::Result<()> {
    if self.show_line_numbers {
      attributes
        .entry("class".into())
        .or_default()
        .push_str(" flex gap-2");
    }
    comrak::html::write_opening_tag(output, "code", attributes)
  }
}

#[derive(Default)]
pub struct HeadingToCAdapter {
  toc: Mutex<Vec<(u8, String, String)>>,
  anchorizer: Mutex<comrak::html::Anchorizer>,
}

impl HeadingToCAdapter {
  pub fn into_toc(self) -> Vec<(u8, String, String)> {
    self.toc.into_inner().unwrap()
  }
}

impl HeadingAdapter for HeadingToCAdapter {
  fn enter(
    &self,
    output: &mut dyn Write,
    heading: &HeadingMeta,
    _sourcepos: Option<Sourcepos>,
  ) -> std::io::Result<()> {
    let mut anchorizer = self.anchorizer.lock().unwrap();

    let anchor = anchorizer.anchorize(heading.content.clone());
    writeln!(output, r#"<h{} id="{anchor}">"#, heading.level)?;

    let mut lock = self.toc.lock().unwrap();
    lock.push((heading.level, heading.content.clone(), anchor));

    Ok(())
  }

  fn exit(
    &self,
    output: &mut dyn Write,
    heading: &HeadingMeta,
  ) -> std::io::Result<()> {
    writeln!(output, "</h{}>", heading.level)?;
    Ok(())
  }
}

#[cfg(feature = "ammonia")]
pub type URLRewriter =
  Arc<dyn (Fn(Option<&ShortPath>, &str) -> String) + Send + Sync>;
