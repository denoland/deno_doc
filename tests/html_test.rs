// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.

use deno_ast::ModuleSpecifier;
use deno_doc::html::pages::SymbolPage;
use deno_doc::html::*;
use deno_doc::DocNode;
use deno_doc::DocParser;
use deno_doc::DocParserOptions;
use deno_graph::source::LoadFuture;
use deno_graph::source::LoadOptions;
use deno_graph::source::LoadResponse;
use deno_graph::source::Loader;
use deno_graph::BuildOptions;
use deno_graph::CapturingModuleAnalyzer;
use deno_graph::GraphKind;
use deno_graph::ModuleGraph;
use futures::future;
use indexmap::IndexMap;
use std::fs;
use std::rc::Rc;

struct SourceFileLoader {}

impl Loader for SourceFileLoader {
  fn load(
    &self,
    specifier: &ModuleSpecifier,
    _options: LoadOptions,
  ) -> LoadFuture {
    let result = if specifier.scheme() == "file" {
      let path = specifier.to_file_path().unwrap();
      fs::read(path)
        .map(|content| {
          Some(LoadResponse::Module {
            specifier: specifier.clone(),
            maybe_headers: None,
            content: content.into(),
          })
        })
        .map_err(|err| err.into())
    } else {
      Ok(None)
    };
    Box::pin(future::ready(result))
  }
}

struct EmptyResolver {}

impl HrefResolver for EmptyResolver {
  fn resolve_path(
    &self,
    current: UrlResolveKind,
    target: UrlResolveKind,
  ) -> String {
    href_path_resolve(current, target)
  }

  fn resolve_global_symbol(&self, _symbol: &[String]) -> Option<String> {
    None
  }

  fn resolve_import_href(
    &self,
    _symbol: &[String],
    _src: &str,
  ) -> Option<String> {
    None
  }

  fn resolve_usage(&self, current_resolve: UrlResolveKind) -> Option<String> {
    current_resolve
      .get_file()
      .map(|current_file| current_file.path.to_string())
  }

  fn resolve_source(&self, _location: &deno_doc::Location) -> Option<String> {
    None
  }
}

async fn get_files(subpath: &str) -> IndexMap<ModuleSpecifier, Vec<DocNode>> {
  let files = fs::read_dir(
    std::env::current_dir()
      .unwrap()
      .join("tests")
      .join("testdata")
      .join(subpath),
  )
  .unwrap();

  let source_files: Vec<ModuleSpecifier> = files
    .into_iter()
    .map(|entry| {
      let entry = entry.unwrap();
      ModuleSpecifier::from_file_path(entry.path()).unwrap()
    })
    .collect();

  let loader = SourceFileLoader {};
  let analyzer = CapturingModuleAnalyzer::default();
  let mut graph = ModuleGraph::new(GraphKind::TypesOnly);
  graph
    .build(
      source_files.clone(),
      &loader,
      BuildOptions {
        module_analyzer: &analyzer,
        ..Default::default()
      },
    )
    .await;

  let parser = DocParser::new(
    &graph,
    &analyzer,
    DocParserOptions {
      diagnostics: false,
      private: false,
    },
  )
  .unwrap();

  let mut source_files = source_files.clone();
  source_files.sort();
  let mut doc_nodes_by_url = IndexMap::with_capacity(source_files.len());
  for source_file in source_files {
    let nodes = parser.parse_with_reexports(&source_file).unwrap();
    doc_nodes_by_url.insert(source_file, nodes);
  }

  doc_nodes_by_url
}

#[tokio::test]
async fn html_doc_files() {
  let files = generate(
    GenerateOptions {
      package_name: None,
      main_entrypoint: None,
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: None,
      composable_output: false,
    },
    get_files("single").await,
  )
  .unwrap();

  let mut file_names = files.keys().collect::<Vec<_>>();
  file_names.sort();

  assert_eq!(
    file_names,
    [
      "./all_symbols.html",
      "./index.html",
      "./~/Bar.html",
      "./~/Bar.prototype.html",
      "./~/Foo.html",
      "./~/Foo.prototype.html",
      "./~/Foobar.html",
      "./~/Foobar.prototype.html",
      "fuse.js",
      "page.css",
      "script.js",
      "search.js",
      "search_index.js",
      "styles.css",
    ]
  );

  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  {
    insta::assert_snapshot!(files.get("./all_symbols.html").unwrap());
    insta::assert_snapshot!(files.get("./index.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("fuse.js").unwrap());
    insta::assert_snapshot!(files.get("page.css").unwrap());
    insta::assert_snapshot!(files.get("script.js").unwrap());
    insta::assert_snapshot!(files.get("search.js").unwrap());
    insta::assert_snapshot!(files.get("search_index.js").unwrap());
  }
}

#[tokio::test]
async fn html_doc_files_rewrite() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");
  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let files = generate(
    GenerateOptions {
      package_name: None,
      main_entrypoint: None,
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    get_files("multiple").await,
  )
  .unwrap();

  let mut file_names = files.keys().collect::<Vec<_>>();
  file_names.sort();

  assert_eq!(
    file_names,
    [
      "./all_symbols.html",
      "./index.html",
      "./~/A.html",
      "./~/A.prototype.html",
      "./~/B.html",
      "./~/B.prototype.html",
      "./~/Bar.html",
      "./~/Bar.prototype.html",
      "./~/Baz.foo.html",
      "./~/Baz.html",
      "./~/Foo.bar.html",
      "./~/Foo.html",
      "./~/Foo.prototype.\"><img src=x onerror=alert(1)>.html",
      "./~/Foo.prototype.foo.html",
      "./~/Foo.prototype.html",
      "./~/Foobar.html",
      "./~/Foobar.prototype.html",
      "./~/Hello.html",
      "./~/Hello.world.html",
      "./~/index.html",
      "foo/~/index.html",
      "foo/~/x.html",
      "fuse.js",
      "page.css",
      "script.js",
      "search.js",
      "search_index.js",
      "styles.css"
    ]
  );

  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  {
    insta::assert_snapshot!(files.get("./all_symbols.html").unwrap());
    insta::assert_snapshot!(files.get("./index.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Baz.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Baz.foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.bar.html").unwrap());
    insta::assert_snapshot!(files
      .get("./~/Foo.prototype.\"><img src=x onerror=alert(1)>.html")
      .unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Hello.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Hello.world.html").unwrap());
    insta::assert_snapshot!(files.get("./~/index.html").unwrap());
    insta::assert_snapshot!(files.get("foo/~/index.html").unwrap());
    insta::assert_snapshot!(files.get("foo/~/x.html").unwrap());
    insta::assert_snapshot!(files.get("fuse.js").unwrap());
    insta::assert_snapshot!(files.get("page.css").unwrap());
    insta::assert_snapshot!(files.get("script.js").unwrap());
    insta::assert_snapshot!(files.get("search.js").unwrap());
    insta::assert_snapshot!(files.get("search_index.js").unwrap());
  }
}

#[tokio::test]
async fn symbol_group() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    Default::default(),
    doc_nodes_by_url,
  )
  .unwrap();

  let mut files = vec![];

  {
    for (short_path, doc_nodes) in &ctx.doc_nodes {
      let symbol_pages =
        generate_symbol_pages_for_module(&ctx, short_path, doc_nodes);

      files.extend(symbol_pages.into_iter().map(
        |symbol_page| match symbol_page {
          SymbolPage::Symbol {
            breadcrumbs_ctx,
            symbol_group_ctx,
            toc_ctx,
          } => {
            let root = ctx.href_resolver.resolve_path(
              UrlResolveKind::Symbol {
                file: short_path,
                symbol: &symbol_group_ctx.name,
              },
              UrlResolveKind::Root,
            );

            let html_head_ctx = pages::HtmlHeadCtx::new(
              &root,
              &symbol_group_ctx.name,
              ctx.package_name.as_ref(),
              Some(short_path),
            );

            Some(pages::SymbolPageCtx {
              html_head_ctx,
              symbol_group_ctx,
              breadcrumbs_ctx,
              toc_ctx,
            })
          }
          SymbolPage::Redirect { .. } => None,
        },
      ));
    }
  }

  let mut files_json = serde_json::to_string_pretty(&files).unwrap();
  files_json.push('\n');

  let testdata_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata");

  #[cfg(feature = "syntect")]
  let symbol_group_json_path = testdata_path.join("symbol_group-syntect.json");
  #[cfg(feature = "tree-sitter")]
  let symbol_group_json_path =
    testdata_path.join("symbol_group-tree-sitter.json");
  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  let symbol_group_json_path = testdata_path.join("symbol_group.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&symbol_group_json_path, &files_json).unwrap();
  }

  let symbol_group_json = fs::read_to_string(symbol_group_json_path).unwrap();

  assert_eq!(files_json, symbol_group_json);
}

#[tokio::test]
async fn symbol_search() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    Default::default(),
    doc_nodes_by_url,
  )
  .unwrap();

  let search_index = generate_search_index(&ctx);
  let mut file_json = serde_json::to_string_pretty(&search_index).unwrap();
  file_json.push('\n');

  let symbol_search_json_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("symbol_search.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&symbol_search_json_path, &file_json).unwrap();
  }

  let symbol_search_json = fs::read_to_string(symbol_search_json_path).unwrap();

  assert_eq!(file_json, symbol_search_json);
}

#[tokio::test]
async fn module_doc() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    FileMode::Single,
    doc_nodes_by_url,
  )
  .unwrap();

  let mut module_docs = vec![];

  for (short_path, doc_nodes) in &ctx.doc_nodes {
    let render_ctx =
      RenderContext::new(&ctx, doc_nodes, UrlResolveKind::File(short_path));
    let module_doc = jsdoc::ModuleDocCtx::new(&render_ctx, short_path);

    module_docs.push(module_doc);
  }

  let mut file_json = serde_json::to_string_pretty(&module_docs).unwrap();
  file_json.push('\n');

  let testdata_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata");

  #[cfg(feature = "syntect")]
  let module_docs_json_path = testdata_path.join("module_doc-syntect.json");
  #[cfg(feature = "tree-sitter")]
  let module_docs_json_path = testdata_path.join("module_doc-tree-sitter.json");
  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  let module_docs_json_path = testdata_path.join("module_doc.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&module_docs_json_path, &file_json).unwrap();
  }

  let module_docs_json = fs::read_to_string(module_docs_json_path).unwrap();
  assert_eq!(file_json, module_docs_json);
}
