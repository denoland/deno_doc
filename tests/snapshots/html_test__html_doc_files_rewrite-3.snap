---
source: tests/html_test.rs
expression: "files.get(\"./~/Bar.html\").unwrap()"
---
<!DOCTYPE html>
<html>
<head>
  <title>Bar - documentation</title>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="doc-current-file" content=".">
  <link rel="stylesheet" href="..&#x2F;styles.css">
  <link rel="stylesheet" href="..&#x2F;page.css">
  <link id="ddocResetStylesheet" rel="stylesheet" href="..&#x2F;reset.css"><script src="..&#x2F;search_index.js" defer></script>
    <script src="..&#x2F;script.js" defer></script>
    <script src="..&#x2F;fuse.js" defer></script>
    <script src="..&#x2F;search.js" defer></script></head>
<body>
<div class="ddoc">
<div><nav class="top-0 sticky bg-white z-50 py-3 h-14" id="topnav">
  <div class="h-full">
    <div><ul class="breadcrumbs"><li><a href="..&#x2F;" class="contextLink">index</a></li><span class="text-[#0F172A]"><svg
	width="16"
	height="16"
	viewBox="0 0 16 16"
	fill="none"
	xmlns="http://www.w3.org/2000/svg">
	<path fill-rule="evenodd" clip-rule="evenodd" d="M5.76748 11.8159C5.5378 11.577 5.54525 11.1972 5.78411 10.9675L8.93431 8L5.78411 5.0325C5.54525 4.80282 5.5378 4.423 5.76748 4.18413C5.99715 3.94527 6.37698 3.93782 6.61584 4.1675L10.2158 7.5675C10.3335 7.68062 10.4 7.83679 10.4 8C10.4 8.16321 10.3335 8.31938 10.2158 8.4325L6.61584 11.8325C6.37698 12.0622 5.99715 12.0547 5.76748 11.8159Z" fill="currentColor" />
</svg>
</span><ul><li>Bar</li></ul></ul>
</div>

      <input
        type="text"
        id="searchbar"
        style="display: none;"
        class="py-2 px-2.5 rounded text-sm border border-gray-300 bg-transparent" />
  </div>
</nav>
<div id="searchResults"></div><div id="content"><main class="symbolGroup" id="symbol_Bar"><article>
      <div>
        <div>
          <div class="text-2xl leading-none break-all">
            <span class="text-Class">class</span>&nbsp;<span class="font-bold">Bar</span>
          </div><div class="symbolSubtitle"><div>
    <span class="type"> extends </span><a class="link" href="..&#x2F;.&#x2F;.&#x2F;~&#x2F;Foo.html">Foo</a><span> </span>
  </div></div></div></div><div><div class="space-y-7" id=""><div class="markdown"><blockquote>
<p>Some quote in bar docs
This quote part is ignored
when getting the title of this doc</p>
</blockquote>
<p>Bar docs</p>
</div></div>
</div>
    </article></main>
<div class="toc">
    <div><div class="usageContent">
    <h3>Usage</h3><div class="markdown"><pre><code>import { Bar } from ".";
</code></pre>
</div></div></div>
  </div></div>
</div>
</div>
</body>
</html>
